kahuna/public/js/crop/controller.js

```diff
@@ -1,236 +1,209 @@
-    };
-    ctrl.broadcastXChange = function (){
+      };
+      ctrl.broadcastXChange = function (){
         $scope.$broadcast('user-x-change', ctrl.inputX);
-    };
-    ctrl.broadcastYChange = function (){
+      };
+      ctrl.broadcastYChange = function (){
         $scope.$broadcast('user-y-change', ctrl.inputY);
-    };
+      };
 
-    //make the view match the ctrl value
-    $scope.$watch(function(){ return ctrl.cropWidth(); }, function(){
+      //make the view match the ctrl value
+      $scope.$watch(function(){ return ctrl.cropWidth(); }, function(){
         ctrl.inputWidth = ctrl.cropWidth();
-    });
-    $scope.$watch(function(){ return ctrl.cropHeight(); }, function(){
+      });
+      $scope.$watch(function(){ return ctrl.cropHeight(); }, function(){
         ctrl.inputHeight = ctrl.cropHeight();
-    });
-    $scope.$watch(function(){ return ctrl.cropX(); }, function(){
+      });
+      $scope.$watch(function(){ return ctrl.cropX(); }, function(){
         ctrl.inputX = ctrl.cropX();
-    });
-    $scope.$watch(function(){ return ctrl.cropY(); }, function(){
+      });
+      $scope.$watch(function(){ return ctrl.cropY(); }, function(){
         ctrl.inputY = ctrl.cropY();
-    });
-
-    ctrl.cropSizeWarning = () => ctrl.cropWidth() < 500;
-
-    ctrl.getRatioString = (aspect) => {
-        if (Number(aspect) === ctrl.landscapeRatio) {
-            return '5:3';
-        } else if (Number(aspect) === ctrl.portraitRatio) {
-            return '4:5';
-        } else if (Number(aspect) === ctrl.squareRatio) {
-            return '1:1';
-        } else if (Number(aspect) === ctrl.videoRatio) {
-            return '16:9';
+      });
+
+      ctrl.cropSizeWarning = () => ctrl.cropWidth() < 500;
+
+      function crop() {
+        // TODO: show crop
+        const coords = {
+          x: Math.round(ctrl.coords.x1),
+          y: Math.round(ctrl.coords.y1),
+          width:  ctrl.cropWidth(),
+          height: ctrl.cropHeight()
+        };
+
+        const ratioString = ctrl.cropOptions.find(_ => _.key === ctrl.cropType).ratioString;
+
+        ctrl.cropping = true;
+
+        mediaCropper.createCrop(ctrl.image, coords, ratioString).then(crop => {
+          // Global notification of action
+          $rootScope.$emit('events:crop-created', {
+            image: ctrl.image,
+            crop: crop
+          });
+
+          $state.go('image', {
+            imageId: imageId,
+            crop: crop.data.id
+          });
+        }).finally(() => {
+          ctrl.cropping = false;
+        });
+      }
+
+      ctrl.callCrop = function() {
+        //prevents return keypress on the crop button posting crop twice
+        if (!ctrl.cropping) {
+          crop();
+        }
+      };
+
+      $scope.$watch('ctrl.cropType', (newCropType, oldCropType) => {
+        const isCropTypeDisabled = ctrl.cropOptions.find(_ => _.key === newCropType).disabled;
+
+        if (isCropTypeDisabled) {
+          ctrl.cropType = oldCropType;
```

Comment:
This is to stop you using keyboard shortcuts to crop to a disabled ratio. That is, if only `video` is enabled, you can press `l` to swap to `landscape` and then crop. That's cheating, so we're now stricter.